### PR TITLE
Create a new kind of revlog entry for reschedule cards on change

### LINKED ISSUE #3357: Create a new revlog entry for "Reschedule cards on change"
Currently, the revlog entry for "Reschedule cards on change" is the same as that for "Set Due Date". This is the cause of several issues, including
- https://github.com/ankitects/anki/issues/2998
- https://github.com/open-spaced-repetition/fsrs4anki-helper/issues/442
- https://github.com/open-spaced-repetition/fsrs4anki-helper/pull/434

There must be other issues too that I am not remembering.

The solution is to use a separate revlog entry to distinguish "Reschedule cards on change" and "Set Due Date".

The first step is to decide the column in the revlog table whose value can be changed while minimizing add-on breakage (they may be making assumptions about the possible values).
Possible options: `type`, `ease`, `factor`

IMO, `type` is the only column where a change would appear to be "logically" correct. However, I am not sure if using `type = 5` for "Reschedule cards on change" would lead to significant add-on breakage.

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
close #3357

### CODE REVIEW

File: rslib/src/scheduler/fsrs/memory_state.rs [hunk 1]
```diff
@@ -249,18 +250,18 @@ fn get_last_revlog_info(revlogs: &[RevlogEntry]) -> HashMap<CardId, LastRevlogIn
...
         .for_each(|(card_id, group)| {
             let mut last_reviewed_at = None;
-            let mut last_revlog_is_manual = false;
+            let mut last_revlog_is_rescheduled = false;
             for e in group.into_iter() {
                 if e.button_chosen >= 1 {
                     last_reviewed_at = Some(e.id.as_secs());
                 }
-                last_revlog_is_manual = e.review_kind == RevlogReviewKind::Manual;
+                last_revlog_is_rescheduled = e.review_kind == RevlogReviewKind::Rescheduled;
```
Comment (dae) [inline rslib/src/scheduler/fsrs/memory_state.rs:258]: What about older cards that were rated in previous Anki versions. Should we be accepting either manual or rescheduled here?

File: rslib/src/stats/graphs/reviews.rs [hunk 2]
```diff
@@ -38,7 +38,7 @@ impl GraphsContext {
                     count.filtered += 1;
                     time.filtered += review.taken_millis;
                 }
-                RevlogReviewKind::Manual => unreachable!(),
+                RevlogReviewKind::Manual | RevlogReviewKind::Rescheduled => unreachable!(),
```
Comment (dae) [inline rslib/src/stats/graphs/reviews.rs:41]: You forgot to exclude the rescheduled case at the top, so it's reachable.

File: rslib/src/scheduler/fsrs/memory_state.rs (hunk 1 above)
Comment (L-M-Sherlock) [reply rslib/src/scheduler/fsrs/memory_state.rs:258]: My concern is the user may feel confused if they set due date  for a card and reschedule all cards, they will only see the first revlog entry generated by `set due date`.

File: rslib/src/stats/graphs/reviews.rs (hunk 2 above)
Comment (L-M-Sherlock) [reply rslib/src/stats/graphs/reviews.rs:43]: aea26e9807f710e3f22aaca9c279d6e4540974ff

File: rslib/src/scheduler/fsrs/memory_state.rs (hunk 1 above)
Comment (dae) [reply rslib/src/scheduler/fsrs/memory_state.rs:258]: This only affects card rescheduling when saving deck options, yes? Won't an upgrading user with previous set due date use see quite different results when rescheduling now, compared to what they were getting previously? Is that intentional?

**Pull request merged**